SRE-2452 rpms: Script to build RPMs from scratch

ryon-jensen opened the pull request:
Script to build all RPMs, including dependencies.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

ryon-jensen committed:
SRE-2452 rpms: Script to build RPMs from scratch

Script to build all RPMs, including dependencies on
RockyLinux 8 and 9.

Introduces a new GHA workflow file for running the script.

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

Doc-only: true

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/5/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/5/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/9/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/20/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/20/display/redirect

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/20/display/redirect

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/20/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/20/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/23/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/24/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/31/display/redirect

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/31/display/redirect

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/31/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/31/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/31/display/redirect

ryon-jensen committed:
Review feedback items

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

Doc-only: true

ryon-jensen committed:
potential fix

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

Doc-only: true

ryon-jensen commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -0,0 +1,142 @@
+dnf_builddep() {
+  local rpms_dst=$1
+  local spec_file="$2"
+  dnf --nogpgcheck --repofrompath rpms,"$rpms_dst" --refresh builddep -y "$spec_file"
```
Updated. I didn't follow your suggestion exactly, but let me know what you think. Thanks!

ryon-jensen committed:
Added support for opensuse/leap

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/43/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15419/43/display/redirect

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -6,52 +6,126 @@ set -uex
+          echo "No build dependencies found in the spec file."
+      fi
+  else
+    dnf --refresh builddep -y "$spec_file"
```
Only concern here is that `--refresh` is going to refresh all of the OS distros which are going to be expensive, NOOP refreshes.  The alternative is to run:
`dnf --disablerepo=\* --enablerepo=local-rpms makecache` either here or in `l_createrepo()` after the `createrepo`, much like you are doing for `zypper`.

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -6,52 +6,126 @@ set -uex
+  zypper install -y -t pattern devel_basis
+
+  # Enable additional repositories if needed
+  if ! zypper lr | grep -q "Main Repository"; then
+      zypper ar -f http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/ "Main Repository"
+  else
+      echo "Main Repository already exists, skipping addition."
+  fi
+  if ! zypper lr | grep -q "Main Update Repository"; then
+      zypper ar -f http://download.opensuse.org/update/leap/$suse_version/oss/ "Main Update Repository"
+  else
+      echo "Main Update Repository already exists, skipping addition."
+  fi
```
Just a bit of DRY here:
```suggestion
  for repo in "Main Repository" "Main Update Repository"; do
    if ! zypper lr | grep -q "$repo"; then
        zypper ar -f http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/ "$repo"
    else
        echo "Main Repository already exists, skipping addition."
    fi
  done
```

ryon-jensen replied:
The URLs need to be different as well, right? I don't know bash as well, is there a way to iterate over a key/value pair? I know, this is my Python mind working. :) 
"Main Repository": http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/
"Main Update Repository": http://download.opensuse.org/update/leap/$suse_version/oss/

brianjmurrell replied:
> The URLs need to be different as well, right?

Ahhh.  I missed that.

> I don't know bash as well, is there a way to iterate over a key/value pair?

Yes, using associative arrays.  Something like:

```bash
declare -A repos
repos["Main Repository"]="distribution/leap/$suse_version/repo/oss
repos["Main Update Repository"]="update/leap/$suse_version/oss"

for repo in "${!repos[@]}"; do
  if ! zypper lr | grep -q "$repo"; then
    zypper ar -f http://download.opensuse.org/"${repos[$repo]}" "$repo"
  else
    echo "$repo already exists, skipping addition."
  fi
done
```

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -6,52 +6,126 @@ set -uex
+      echo "Main Update Repository already exists, skipping addition."
+  fi
+
+  # Ignore dpdk because want to use RPMs built from daos-stack/dpdk
```
Is the `dpdk` part of the comment still relevant?

ryon-jensen replied:
Thanks, will remove.

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -6,52 +6,126 @@ set -uex
+      rpmdevtools \
+      createrepo_c \
+      make \
+      python3-Sphinx
```
Any idea what are requiring this?  If it's failing `rpmbuild`s that's a packaging bug.

ryon-jensen replied:
There's a comment for python3-Sphinx:
  # python3-Sphinx is needed by dpdk. The file /usr/bin/sphinx-build is listed as a dependency in the spec file,
  #   however, zypper isn't able to resolve a filename to a package and the scripting to handle that case is pretty
  #   messy, especially because "zypper what-provides" actually returns three options for sphinx-build. So, forgoing
  #   automation here and simply hardcoding the dependency.
I recognize it's not ideal, but weighing the effort and complexity (at least for me) for managing the dependency dynamically compared to hardcoding it and dealing with the impact on maintainability, I chose this approach. Definitely  open to suggestions.

brianjmurrell replied:
So that `dpdk` needs it means that the `BuildRequires:` in `dpdk.spec` should specify it.  Did you get a failure to build `dpdk` without it?  If you had a link to the failure that'd be awesome.

ryon-jensen replied:
`BuildRequires:` specifies `/usr/bin/sphinx-build`, however, zypper doesn't resolve filenames to packages to install.

brianjmurrell replied:
The other approach is that even if you want to continue to use `zypper` for the bulk of this script, recognize that `zypper` is inferior (or we are just ignorant) in regards to installing build-dependencies and install and use `dnf` for that.  Perhaps if we are more obvious about this short-coming (or our ignorance) in `zypper`, somebody with more knowledge of `zypper` that is using and looking at this script can find/suggest a better way.

But I think by simply adding `python3-Sphinx` as an exception to the package install list here we are making the issue much (much!) more subtle than if we were to use `dnf` with a nice big comment about why we are using it.

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -6,52 +6,126 @@ set -uex
       createrepo \
+      dnf-plugins-core \
       epel-release \
       make # shouldn't be needed here, but dependency spec files don't include
```
Actually it is required here since you are using `make rpms` which is independent of any requirement for `make` to build an RPM.

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -60,15 +134,62 @@ setup_rpm_build_env() {
+
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+    zypper refresh
+  fi
```
As mentioned earlier, you could use `dnf … makecache` here instead of using `--refresh` elsewhere.

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -60,15 +134,62 @@ setup_rpm_build_env() {
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+      prepare_env_leap
+    else
+      prepare_env_rocky
```
```suggestion
      prepare_env_el
```

perhaps?

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -60,15 +134,62 @@ setup_rpm_build_env() {
+}
+
+verify_os_version() {
+  $OS_ID
```
I think this would be a syntax error here.  Maybe you meant:

```suggestion
  echo "$OS_ID"

brianjmurrell commented on utils/build_rpms_from_scratch.sh:
```diff
@@ -0,0 +1,261 @@
+    git clone https://github.com/daos-stack/"$repo_name".git
+    cd "$repo_name"
+      install_dependencies "$repo_name".spec
+      make rpms RPM_BUILD_OPTIONS="--nocheck" # --nocheck = don't run tests during rpm build process
```
```suggestion
      MAKEARGS=()
      if [ -f Makefile-rpm.mk ]; then
        MAKEARGS=(-f Makefile-rpm.mk)
      fi
      if [ -d utils/rpms ]; then
        MAKEARGS+=(-C utils/rpms)
      fi
      make "${MAKEARGS[@]}" rpms RPM_BUILD_OPTIONS="--nocheck" # --nocheck = don't run tests during rpm build process
```
will let you use this for _raft_ and _daos_ also and be future-proof against any other project that utilizes these exceptions.

ryon-jensen replied:
I think it's a little more complicated than this. build_and_copy_rpm also clones the repo and because raft is a submodule of daos, the daos repo must be cloned first, then raft rpms built, then daos rpms can be built. So a bit of a chicken and egg issue with the way I wrote build_and_copy_rpms. I appreciate the help trying to remove the redundancy, though. I've probably already spent too much time trying to figure out how to remove the duplicate code without making things too complicated.

brianjmurrell replied:
> I think it's a little more complicated than this.

I don't think it is.

> build_and_copy_rpm also clones the repo and because raft is a submodule of daos, the daos repo must be cloned first, then raft rpms built, then daos rpms can be built.

You can clone and build the raft repo independently of daos just like you do the other dozen repos.  There is no need to clone daos and build raft out of the daos clone.

ryon-jensen replied:
My concern is getting the correct raft commit that's actually used for daos. The 'fun' of it being a submodule is it doesn't have to be the tip of master. Right now it looks like the submodule daos uses is 1 commit behind the tip of master.

brianjmurrell replied:
Yes, this is bad and I have said so any number of times to the developers of raft.  Our dependency policy is that we always use the tip of master of all of our dependencies.  That is how our users know what to use while we do not unnecessarily (IMO) complicate our dependency handling with tags and branches.

Commits in any dependency that are not ready for use should be in branches/PRs, just like we do for every other project (esp. daos).  We have all of the capabilities that we need to do any kind of development or testing in CI from branches/PRs so there is no reason to be pushing things to a project's master that we don't intend to be what a production (or current development) site to use.

We need to stop allowing raft to be treated exceptionally in this process/policy as it only serves to confuse and make us have to have exceptions to policies with no good reason.

ryon-jensen committed:
Attempting not failing if docker image doesn't exist

Doc-only: true

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

brianjmurrell commented on .github/workflows/test_rpm_from_scratch_process.yml:
```diff
@@ -0,0 +1,48 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: ["rockylinux:8.6",
+             "rockylinux:8.8",
+             "rockylinux:8.10",
+             "rockylinux:9.2",
+             "rockylinux:9.3",
+             "opensuse/leap:15.4",
+             "opensuse/leap:15.5",
+             "opensuse/leap:15.6"]
```
I wonder if using our self-hosted runner and therefore our own docker hub/cache, you will find more of these older versions.  @JohnMalmberg Would we have these old point releases in our docker runners?

JohnMalmberg replied:
On the docker runners we are only keeping rockylinux8 and 9 up to date, no dot releases.
We have opensuse/leap:15.[2-5] and opensuse/leap-dnf:15.[4-5]

Just updated so that soon they will have leap and leap-dnf 15.6 and Fedora:41

brianjmurrell replied:
> On the docker runners we are only keeping rockylinux8 and 9 up to date, no dot releases.

This is a good case why it would be useful if we were keeping the EL point releases in our docker hub.

ryon-jensen committed:
Other feedback and fixes

Doc-only: true
Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

ryon-jensen committed:
Other feedback and fixes

- DRY-ed some stuff
- Fixed use of wrong variable
- Fixed miss of directory change

Doc-only: true
Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

ryon-jensen committed:
Added more OS checking in script

Doc-only: true
Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

ryon-jensen committed:
Fixes

- matching of el major version for Power Tools repo was wrong
- Added fail-fast: false (at least for now)

Doc-only: true
Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

ryon-jensen committed:
Fixes

- Not sure why I was using actions/checkout@v3 - bumped to v4
- Removed leap 15.6 for now. Having issues with it building
- A little reorganization
- Removed accidental add of file

Doc-only: true
Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>
